File: 484-find-permutation/484-find-permutation.py
class Solution:
    def findPermutation(self, s: str) -> List[int]:
        n, res, stack= len(s), [], []
        for i in range(n):
            if s[i] == 'I':
                stack.append(i + 1)
                while stack:
                    res.append(stack.pop())
            else:
                stack.append(i + 1)
        stack.append(n + 1)
        while stack:
            res.append(stack.pop())
        return res

File: 484-find-permutation/484-find-permutation.java
class Solution {
    public int[] findPermutation(String s) {
        int[] res = new int[s.length() + 1];
        Deque<Integer> stack = new ArrayDeque<>();
        int j = 0;
        for(int i = 1; i <= s.length(); i++) {
            stack.push(i);
            if(s.charAt(i - 1) == 'I') {
                while(!stack.isEmpty()) {
                    res[j++] = stack.pop();
                }
            }
        }
        stack.push(s.length() + 1);
        while(!stack.isEmpty()) res[j++] = stack.pop();
        return res;
    }
}